app/src/main/java/com/hover/stax/requests/Request.kt

```diff
@@ -0,0 +1,225 @@
+	}
+
+	private fun getDescription(contact: StaxContact?, c: Context): String {
+		return c.getString(R.string.descrip_request, contact!!.shortName() ?: "")
+	}
+
+	fun generateRecipientString(contacts: List<StaxContact?>): String {
+		val phones = StringBuilder()
+		for (r in contacts.indices) {
+			if (phones.isNotEmpty()) phones.append(",")
+			contacts[r]?.let { phones.append(it.accountNumber) }
+		}
+		return phones.toString()
+	}
+
+	fun generateWhatsappRecipientString(contacts: List<StaxContact?>?, c: Channel?): String {
+		val phones = StringBuilder()
+		if(contacts !=null) {
+			for (r in contacts.indices) {
+				if (phones.isNotEmpty()) phones.append(",")
+				contacts[r]?.let{
+					phones.append(PhoneHelper.getInternationalNumberNoPlus(it.accountNumber,
+						c?.countryAlpha2 ?: Lingver.getInstance().getLocale().country))
+				}
+			}
+		}
+		return phones.toString()
+	}
+
+	fun generateMessage(c: Context): String {
+		val amountString = if (amount != null) c.getString(R.string.sms_amount_detail,
+			formatAmount(amount!!))
+		else ""
+		val noteString =
+			if (note != null && !TextUtils.isEmpty(note)) c.getString(R.string.sms_note_detail,
+				note)
+			else " "
+		val paymentLink = generateStaxLink(c)
+		return c.getString(R.string.sms_request_template, amountString, noteString, paymentLink)
+	}
+
+	private fun generateStaxLink(c: Context): String {
+		val amountNoFormat = if (amount != null && !amount!!.isEmpty()) amount!! else "0.00"
+		val requesterNumber = requester_number!!.replace("+", "")
+		val params = c.getString(R.string.payment_url_end,
+			amountNoFormat,
+			requester_institution_id,
+			requesterNumber,
+			now())
+		Timber.i("encrypting from: %s", params)
+		val encryptedString = encryptBijective(params, c)
+		Timber.i("link: %s", c.resources.getString(R.string.payment_root_url, encryptedString))
+		return c.resources.getString(R.string.payment_root_url, encryptedString)
+	}
+
+	private fun encryptBijective(value: String, c: Context): String {
+		val valueChar = value.toCharArray()
+		val result = CharArray(value.toCharArray().size)
+		for (i in valueChar.indices) {
+			when (valueChar[i]) {
+				'.' -> result[i] = bijectiveExt(c)[0]
+				'-' -> result[i] = bijectiveExt(c)[1]
+				',' -> result[i] = bijectiveExt(c)[2]
+				else -> {
+					val valueToInteger = valueChar[i].toString().toInt()
+					result[i] = bijectiveKey(c)[valueToInteger]
+				}
+			}
+		}
+		return String(result)
+	}
+
+	override fun toString(): String {
+		return if (description == null) "Request from $requester_number" else description!!
+	}
+
+	companion object {
+		const val PAYMENT_LINK_SEPERATOR = "-"
+		private const val TAG = "Request"
+		fun isShortLink(link: String): Boolean {
+			return link.length <= 25
+		}
+
+		val encryptionSettings: Encryption.Builder
+			get() = Encryption.Builder().setKeyLength(128).setKeyAlgorithm("AES")
+				.setCharsetName("UTF8").setIterationCount(65536).setKey("ves€Z€xs€aBKgh")
```

Comment:
Should we be passing this here? could we make it a variable? @davkutalek thoughts?

Reply:
@johnswekenyi yeah, there are serious issues with this current approach. We did it this way to work offline and have been planning to re-do it but waiting since planning full redesign. Lumi has been working on the new designs for us, so we should get to it soon-ish.

Reply:
I'm generally not a big fan of having functions inside model classes other than constructors just to keep them clean. Functionality can be moved out to a view model or helper class. Thoughts @johnswekenyi @davkutalek

Reply:
@alexosugo the encryption could be moved to a helper class so that the request object doesn't have to worry about how the encryption is done, just calls methods to encrypt/decrypt. It is fundamental to the request object, so it should not be in a view model since those should deal with how the app works functionally from the user's point of view.

Reply:
@davkutalek Agreed. However, the generate methods can be moved out. They don't impact the object in any way.